src/convnets/resnets/core.jl

```diff
@@ -0,0 +1,342 @@
+                                              reduction_factor, activation,
+                                              norm_layer, prenorm, attn_fn)
+end
+
+function template_builder(downsample_fn::Union{typeof(downsample_conv),
+                                               typeof(downsample_pool),
+                                               typeof(downsample_identity)};
+                          norm_layer = BatchNorm, prenorm = false)
+    return (args...; kwargs...) -> downsample_fn(args...; kwargs..., norm_layer, prenorm)
+end
+
+function configure_block(block_template, layers::Vector{Int}; expansion,
+                         downsample_templates::Vector, inplanes::Integer = 64,
+                         drop_path_rate = 0.0, drop_block_rate = 0.0, kargs...)
+    pathschedule = linear_scheduler(drop_path_rate; depth = sum(layers))
+    blockschedule = linear_scheduler(drop_block_rate; depth = sum(layers))
+    # closure over `idxs`
+    function get_layers(idxs::NTuple{2, Int})
+        stage_idx, block_idx = idxs
+        planes = 64 * 2^(stage_idx - 1)
+        stride = get_stride(idxs)
+        downsample_fns = downsample_templates[stage_idx]
+        downsample_fn = (stride != 1 || inplanes != planes * expansion) ?
+                        downsample_fns[1] :
+                        downsample_fns[2]
+        # DropBlock, DropPath both take in rates based on a linear scaling schedule
+        schedule_idx = sum(layers[1:(stage_idx - 1)]) + block_idx
+        drop_path = DropPath(pathschedule[schedule_idx])
+        drop_block = DropBlock(blockschedule[schedule_idx])
+        block = block_template(inplanes, planes; stride, drop_path, drop_block)
+        downsample = downsample_fn(inplanes, planes * expansion; stride)
+        # inplanes increases by expansion after each block
+        inplanes = (planes * expansion)
+        return ((block, downsample), inplanes)
+    end
+    return get_layers
+end
+
+# Makes the main stages of the ResNet model. This is an internal function and should not be 
+# used by end-users. `block_fn` is a function that returns a single block of the ResNet. 
+# See `basicblock` and `bottleneck` for examples. A block must define a function 
+# `expansion(::typeof(block))` that returns the expansion factor of the block.
+function resnet_stages(block_fn, block_repeats::Vector{Int}, inplanes::Integer;
+                       downsample_vec::Vector, connection = addact,
+                       activation = relu, kwargs...)
+    outplanes = 0
+    # Configure block template
+    block_template = template_builder(block_fn; kwargs...)
+    downsample_templates = map(x -> template_builder.(x), downsample_vec)
+    get_layers = configure_block(block_template, block_repeats; inplanes,
+                                 downsample_templates,
+                                 expansion = expansion_factor(block_fn), kwargs...)
```

Comment:
To my understanding, part of the reason for the double higher-order function pattern was so `resnet_stages` could take in `get_layers` directly as the first argument instead of constructing that itself. Did you try this and/or run into any challenges with it?

Reply:
Do you mean constructing this inside of `resnet` and passing it in? We could but we increase the number of kwargs in `resnet`. Though given that everything is going to be documented I don't think it should make a difference where it is - I'm okay either way

Reply:
I'm not sure I follow, isn't the current set of kwargs in `resnet_stages` either passed through from `resnet` or calculated in `resnet`? So it's almost like a strict subset, if anything.

Reply:
Yep, it's exactly the same. I meant to say that we will have to write it out explicitly in the `resnet` function signature. It doesn't make much of a difference honestly (maybe `resnet` looks a tad bit more intimidating). I'll shift that part to happen inside of `resnet`